### PR TITLE
optimizer-hints, correlated-subquery-optimization: add docs for `no_decorrelate` hint

### PR DESCRIPTION
**Pull request by time-and-fate**
### What is changed, added or deleted? (Required)

Add description for the new `no_decorrelate` hint, and update some descriptions for the correlated subquery.

This is expected for v6.4.0.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [ ] v6.3 (TiDB 6.3 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s): https://github.com/pingcap/tidb/pull/37633

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: correlated-subquery-optimization.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ TiDB 之所以要进行这样的改写，是因为关联子查询每次子查询
 
 这种改写的弊端在于，在关联没有被解除时，优化器是可以使用关联列上的索引的。也就是说，虽然这个子查询可能被重复执行多次，但是每次都可以使用索引过滤数据。而解除关联的变换上，通常是会导致关联列的位置发生改变而导致虽然子查询只被执行了一次，但是单次执行的时间会比没有解除关联时的单次执行时间长。
 
-因此，在外部的值比较少的情况下，不解除关联依赖反而可能对执行性能更有帮助。这时可以通过[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中关闭`子查询去关联`优化规则的方式来关闭这个优化。
+因此，在外部的值比较少的情况下，不解除关联依赖反而可能对执行性能更有帮助。这时可以通过使用 Hint [`NO_DECORRELATE`](/optimizer-hints.md#no_decorrelate) 或[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中关闭`子查询去关联`优化规则的方式来关闭这个优化。
```
Comment (Reminiscent) [inline correlated-subquery-optimization.md:21]: 要不要提一下推荐使用 Hint，而不是后一种方式。

File: optimizer-hints.md [hunk 2]
```diff
@@ -191,6 +191,68 @@ EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 FROM t
 
 在上述例子中可以看到，在使用了 Hint 之后，TiDB 可以选择由表 `t1` 作为驱动表的 IndexJoin 的执行方式。
 
+### NO_DECORRELATE()
+
+`NO_DECORRELATE()` 提示优化器不要尝试对指定查询块对应的子查询解除关联。该 Hint 作用于包含关联列的 `EXISTS`、`IN`、`ANY/ALL/SOME` 和标量子查询，即通常所说的关联子查询。
+
+将该 Hint 写在一个查询块中时，该子查询和其外部查询块之间会保持关联列，使用 Apply 算子来执行。
```
Comment (Reminiscent) [inline optimizer-hints.md:198]: “会保持关联列” 感觉有点奇怪

**Review by Reminiscent — APPROVED** [review]
Rest LGTM

File: correlated-subquery-optimization.md (hunk 1 above)
Comment (time-and-fate) [reply correlated-subquery-optimization.md]: 已更新。

File: optimizer-hints.md (hunk 2 above)
Comment (time-and-fate) [reply optimizer-hints.md]: 已更新。

**Review by Reminiscent — APPROVED** [review]

File: correlated-subquery-optimization.md [hunk 3]
```diff
@@ -18,7 +18,7 @@ TiDB 之所以要进行这样的改写，是因为关联子查询每次子查询
 
 这种改写的弊端在于，在关联没有被解除时，优化器是可以使用关联列上的索引的。也就是说，虽然这个子查询可能被重复执行多次，但是每次都可以使用索引过滤数据。而解除关联的变换上，通常是会导致关联列的位置发生改变而导致虽然子查询只被执行了一次，但是单次执行的时间会比没有解除关联时的单次执行时间长。
 
-因此，在外部的值比较少的情况下，不解除关联依赖反而可能对执行性能更有帮助。这时可以通过[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中关闭`子查询去关联`优化规则的方式来关闭这个优化。
+因此，在外部的值比较少的情况下，不解除关联依赖反而可能对执行性能更有帮助。这时可以通过使用 Hint [`NO_DECORRELATE`](/optimizer-hints.md#no_decorrelate) 或[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中关闭`子查询去关联`优化规则的方式来关闭这个优化。在一般情况下，使用 Hint 并在需要时配合[执行计划管理](/sql-plan-management.md)功能是更推荐的方式。
```
Comment (TomShawn) [inline correlated-subquery-optimization.md:21]:
```suggestion
因此，在外部的值比较少的情况下，不解除关联依赖反而可能对执行性能更有帮助。这时可以通过使用 Optimizer Hint [`NO_DECORRELATE`](/optimizer-hints.md#no_decorrelate) 或[优化规则及表达式下推的黑名单](/blocklist-control-plan.md)中关闭“子查询去关联”优化规则的方式来关闭这个优化。在一般情况下，推荐使用 Optimizer Hint 并在需要时配合[执行计划管理](/sql-plan-management.md)功能来解除关联。
```

File: correlated-subquery-optimization.md [hunk 4]
```diff
@@ -49,7 +49,32 @@ explain select * from t1 where t1.a < (select sum(t2.a) from t2 where t2.b = t1.
 
 上面是优化生效的情况，可以看到 `HashJoin_11` 是一个普通的 `inner join`。
 
-接下来，关闭关联规则：
+接下来，通过 Hint 提示优化器不要对该子查询解除关联：
```
Comment (TomShawn) [inline correlated-subquery-optimization.md:52]:
```suggestion
接下来，通过Optimizer Hint `NO_DECORRELATE` 提示优化器不对该子查询解除关联：
```

File: correlated-subquery-optimization.md [hunk 5]
```diff
@@ -49,7 +49,32 @@ explain select * from t1 where t1.a < (select sum(t2.a) from t2 where t2.b = t1.
...
+
+{{< copyable "sql" >}}
+
+```sql
+explain select * from t1 where t1.a < (select /*+ NO_DECORRELATE() */ sum(t2.a) from t2 where t2.b = t1.b);
+```
+
+```
```
Comment (TomShawn) [inline correlated-subquery-optimization.md:60]:
````suggestion
```sql
````

File: optimizer-hints.md [hunk 6]
```diff
@@ -191,6 +191,68 @@ EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 FROM t
 
 在上述例子中可以看到，在使用了 Hint 之后，TiDB 可以选择由表 `t1` 作为驱动表的 IndexJoin 的执行方式。
 
+### NO_DECORRELATE()
+
+`NO_DECORRELATE()` 提示优化器不要尝试对指定查询块对应的子查询解除关联。该 Hint 作用于包含关联列的 `EXISTS`、`IN`、`ANY/ALL/SOME` 和标量子查询，即通常所说的关联子查询。
```
Comment (TomShawn) [inline optimizer-hints.md:196]:
```suggestion
`NO_DECORRELATE()` 提示优化器不要尝试解除指定查询块中对应子查询的关联。该 Hint 适用于包含关联列的 `EXISTS`、`IN`、`ANY`、`ALL`、`SOME` 和标量子查询，即关联子查询。
```

File: optimizer-hints.md [hunk 7]
```diff
@@ -191,6 +191,68 @@ EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 FROM t
...
+
+将该 Hint 写在一个查询块中时，对于该子查询和其外部查询块之间的关联列，优化器将不再尝试解关联，而是始终使用 Apply 算子来执行。
```
Comment (TomShawn) [inline optimizer-hints.md:198]:
```suggestion
将该 Hint 写在一个查询块中后，对于该子查询和其外部查询块之间的关联列，优化器将不再尝试解除关联，而是始终使用 Apply 算子来执行查询。
```

File: correlated-subquery-optimization.md (hunk 3 above)
Comment (time-and-fate) [reply correlated-subquery-optimization.md:21]: 这个不对。是使用这种方式来关闭这个优化，也就是禁用解关联。

File: optimizer-hints.md [hunk 8]
```diff
@@ -191,6 +191,68 @@ EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 FROM t
...
+```
+
+{{< copyable "sql" >}}
+
+```sql
+-- 不使用 NO_DECORRELATE()
+explain select * from t1 where t1.a < (select sum(t2.a) from t2 where t2.b = t1.b);
+```
+
+```
```
Comment (TomShawn) [inline optimizer-hints.md:216]:
````suggestion
```sql
````

File: optimizer-hints.md [hunk 9]
```diff
@@ -191,6 +191,68 @@ EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 FROM t
...
+
+{{< copyable "sql" >}}
+
+```sql
+-- 使用 NO_DECORRELATE()
+explain select * from t1 where t1.a < (select /*+ NO_DECORRELATE() */ sum(t2.a) from t2 where t2.b = t1.b);
+```
+
+```
```
Comment (TomShawn) [inline optimizer-hints.md:239]:
````suggestion
```sql
````

File: correlated-subquery-optimization.md (hunk 3 above)
Comment (time-and-fate) [reply correlated-subquery-optimization.md:21]: 这里是根据上面的 comments 写的。这里的意思是在列举的两种方式中，一般推荐用 hint 而不是使用优化规则黑名单。

File: optimizer-hints.md [hunk 10]
```diff
@@ -191,6 +191,68 @@ EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 FROM t
...
+| │     └─Selection_22             | 9990.00  | cop[tikv] |               | not(isnull(test.t2.b))                                                                                       |
+| │       └─TableFullScan_21       | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                                               |
+| └─TableReader_15(Probe)          | 9990.00  | root      |               | data:Selection_14                                                                                            |
+|   └─Selection_14                 | 9990.00  | cop[tikv] |               | not(isnull(test.t1.b))                                                                                       |
+|     └─TableFullScan_13           | 10000.00 | cop[tikv] | table:t1      | keep order:false, stats:pseudo                                                                               |
++----------------------------------+----------+-----------+---------------+--------------------------------------------------------------------------------------------------------------+
+```
+
```
Comment (TomShawn) [inline optimizer-hints.md:231]: 是否可以简要说一下从执行结果中哪里可以看出是否已解除关联呢？

File: correlated-subquery-optimization.md (hunk 3 above)
Comment (time-and-fate) [reply correlated-subquery-optimization.md:21]: 其它的变动我没问题了~

Comment (TomShawn) [reply correlated-subquery-optimization.md:21]: 已修改为“禁止解除关联”

File: optimizer-hints.md (hunk 10 above)
Comment (time-and-fate) [reply optimizer-hints.md:231]: 已更新~

**Review by TomShawn — APPROVED** [review]
LGTM

File: optimizer-hints.md [hunk 11]
```diff
@@ -191,6 +191,72 @@ EXPLAIN SELECT * FROM t WHERE EXISTS (SELECT /*+ SEMI_JOIN_REWRITE() */ 1 FROM t
 
 在上述例子中可以看到，在使用了 Hint 之后，TiDB 可以选择由表 `t1` 作为驱动表的 IndexJoin 的执行方式。
 
+### NO_DECORRELATE()
+
+`NO_DECORRELATE()` 提示优化器不要尝试解除指定查询块中对应子查询的关联。该 Hint 适用于包含关联列的 `EXISTS`、`IN`、`ANY`、`ALL`、`SOME` 和标量子查询，即关联子查询。
+
+将该 Hint 写在一个查询块中后，对于该子查询和其外部查询块之间的关联列，优化器将不再尝试解除关联，而是始终使用 Apply 算子来执行查询。
+
+默认情况下，TiDB 会尝试对关联子查询[解除关联](/correlated-subquery-optimization.md)，以达到更高的执行效率。但是在[一部分场景](/correlated-subquery-optimization.md#限制)下，解除关联会让执行效率变得更低。这种情况下，可以使用该 Hint 来人工提示优化器不要进行解除关联操作。例如：
```
Comment (hfxsd) [inline optimizer-hints.md:200]:
```suggestion
默认情况下，TiDB 会尝试对关联子查询[解除关联](/correlated-subquery-optimization.md)，以达到更高的执行效率。但是在[一部分场景](/correlated-subquery-optimization.md#限制)下，解除关联反而会降低执行效率。这种情况下，可以使用该 Hint 来人工提示优化器不要进行解除关联操作。例如：
```

**Review by hfxsd — APPROVED** [review]

**Pull request merged**